File: src/lfmcmc.cpp
#include "cpp11.hpp"
#include "cpp11/external_pointer.hpp"
#include "cpp11/r_vector.hpp"
#include "cpp11/sexp.hpp"
#include "cpp11/doubles.hpp"
#include <iostream>

#include "epiworld-common.h"

using namespace epiworld;

#define TData_default std::vector< double >

#define WrapLFMCMC(a) \
    cpp11::external_pointer<LFMCMC<TData_default>> (a)

inline cpp11::sexp lfmcmc_as_sexp(LFMCMC<TData_default> * lfmcmc) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc, false);
    cpp11::sexp lfmcmc_ptr_s(lfmcmc_ptr);
    lfmcmc_ptr_s.attr("class") = "epiworld_lfmcmc";
    return lfmcmc_ptr_s;
}

// LFMCMC definitions:
// https://github.com/UofUEpiBio/epiworld/tree/master/include/epiworld/math/lfmcmc

// *************************************
// LFMCMC Function
// *************************************

[[cpp11::register]]
SEXP LFMCMC_cpp(
    SEXP model
) {
    WrapLFMCMC(lfmcmc_ptr)(
        new LFMCMC<TData_default>()
    );

    if (Rf_inherits(model, "epiworld_model")) {
        cpp11::external_pointer<Model<int>> modelptr(model);
        lfmcmc_ptr->set_rand_engine(modelptr->get_rand_endgine());
    } else {
        auto new_ptr = std::make_shared<std::mt19937>(std::mt19937());
        lfmcmc_ptr->set_rand_engine(new_ptr);
    }

    return lfmcmc_ptr;
}

// *************************************
// LFMCMC Run Function
// *************************************

[[cpp11::register]]
SEXP run_lfmcmc_cpp(
    SEXP lfmcmc,
    std::vector<double> params_init_,
    size_t n_samples_,
    double epsilon_,
    int seed
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->run(params_init_, n_samples_, epsilon_, seed);
    return lfmcmc;
}

// *************************************
// LFMCMC Setup Functions
// *************************************

// observed_data_ should be of type TData
[[cpp11::register]]
SEXP set_observed_data_cpp(
    SEXP lfmcmc,
    std::vector< double > observed_data_
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->set_observed_data(observed_data_);
    return lfmcmc;
}

[[cpp11::register]]
SEXP set_proposal_fun_cpp(
    SEXP lfmcmc,
    cpp11::function fun
) {

    LFMCMCProposalFun<TData_default> fun_call = [fun](
        std::vector< epiworld_double >& new_params,
        const std::vector< epiworld_double >& old_params,
        LFMCMC<TData_default> * lfmcmc_obj
        ) -> void {

        auto params_doubles = cpp11::doubles(old_params);
        auto lfmcmc_ptr = lfmcmc_as_sexp(lfmcmc_obj);
        auto res_tmp = cpp11::doubles(fun(params_doubles, lfmcmc_ptr));

        std::copy(
            res_tmp.begin(),
            res_tmp.end(),
            new_params.begin()
            );

        return;
    };

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);

    lfmcmc_ptr->set_proposal_fun(fun_call);

    return lfmcmc;
}

// Use proposal function defined in epiworld
[[cpp11::register]]
SEXP use_proposal_norm_reflective_cpp(
    SEXP lfmcmc
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->set_proposal_fun(make_proposal_norm_reflective<std::vector<double>>(.5, 0, 1));
    return lfmcmc;
}

[[cpp11::register]]
SEXP set_simulation_fun_cpp(
    SEXP lfmcmc,
    cpp11::function fun
) {

    LFMCMCSimFun<TData_default> fun_call = [fun](
        const std::vector<epiworld_double>& params,
        LFMCMC<TData_default> * lfmcmc_obj
        ) -> TData_default {

        auto params_doubles = cpp11::doubles(params);
        auto lfmcmc_ptr = lfmcmc_as_sexp(lfmcmc_obj);
        return cpp11::as_cpp<TData_default>(
            cpp11::doubles(fun(params_doubles, lfmcmc_ptr))
            );
    };

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);

    lfmcmc_ptr->set_simulation_fun(fun_call);

    return lfmcmc;
}

[[cpp11::register]]
SEXP set_summary_fun_cpp(
    SEXP lfmcmc,
    cpp11::function fun
) {

    LFMCMCSummaryFun<TData_default> fun_call = [fun](
        std::vector< epiworld_double >& res,
        const TData_default& dat,
        LFMCMC<TData_default>* lfmcmc_obj
        ) -> void {

        auto dat_int = cpp11::doubles(dat);
        auto lfmcmc_ptr = lfmcmc_as_sexp(lfmcmc_obj);
        auto res_tmp = cpp11::doubles(fun(dat_int, lfmcmc_ptr));


        if (res.size() == 0u)
            res.resize(res_tmp.size());

        std::copy(res_tmp.begin(), res_tmp.end(), res.begin());

        return;
    };

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);

    lfmcmc_ptr->set_summary_fun(fun_call);

    return lfmcmc;
}

[[cpp11::register]]
SEXP set_kernel_fun_cpp(
    SEXP lfmcmc,
    cpp11::function fun
) {

    LFMCMCKernelFun<TData_default> fun_call = [fun](
        const std::vector< epiworld_double >& simulated_stats,
        const std::vector< epiworld_double >& observed_stats,
        epiworld_double epsilon,
        LFMCMC<TData_default> * lfmcmc_obj
        ) -> epiworld_double {

        auto sim_stats_doubles = cpp11::doubles(simulated_stats);
        auto obs_stats_doubles = cpp11::doubles(observed_stats);

        auto lfmcmc_ptr = lfmcmc_as_sexp(lfmcmc_obj);

        return cpp11::as_cpp<epiworld_double>(
            fun(sim_stats_doubles, obs_stats_doubles, epsilon, lfmcmc_ptr)
            );
    };

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);

    lfmcmc_ptr->set_kernel_fun(fun_call);

    return lfmcmc;
}

// Use kernel function defined in epiworld
[[cpp11::register]]
SEXP use_kernel_fun_gaussian_cpp(
    SEXP lfmcmc
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->set_kernel_fun(kernel_fun_gaussian<std::vector<double>>);
    return lfmcmc;
}

// *************************************
// LFMCMC Getters - Params and Stats
// *************************************

[[cpp11::register]]
cpp11::writable::doubles get_mean_params_cpp(
    SEXP lfmcmc
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_mean_params());
}

[[cpp11::register]]
cpp11::writable::doubles get_mean_stats_cpp(
    SEXP lfmcmc
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_mean_stats());
}

[[cpp11::register]]
cpp11::writable::doubles get_initial_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_initial_params());

}

[[cpp11::register]]
cpp11::writable::doubles get_current_proposed_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_current_proposed_params());

}

[[cpp11::register]]
cpp11::writable::doubles get_current_accepted_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_current_accepted_params());

}

[[cpp11::register]]
cpp11::writable::doubles get_current_proposed_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_current_proposed_stats());

}

[[cpp11::register]]
cpp11::writable::doubles get_current_accepted_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_current_accepted_stats());

}

[[cpp11::register]]
cpp11::writable::doubles get_observed_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_observed_stats());

}

[[cpp11::register]]
SEXP get_all_sample_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_sample_params());

}

[[cpp11::register]]
SEXP get_all_sample_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_sample_stats());

}

[[cpp11::register]]
cpp11::writable::doubles get_all_sample_acceptance_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_sample_acceptance());

}

[[cpp11::register]]
cpp11::writable::doubles get_all_sample_drawn_prob_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_sample_drawn_prob());

}

[[cpp11::register]]
cpp11::writable::doubles get_all_sample_kernel_scores_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_sample_kernel_scores());

}

[[cpp11::register]]
SEXP get_all_accepted_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_accepted_params());

}

[[cpp11::register]]
SEXP get_all_accepted_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_accepted_stats());

}

[[cpp11::register]]
SEXP get_all_accepted_kernel_scores_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return cpp11::doubles(lfmcmc_ptr->get_all_accepted_kernel_scores());

}

// *************************************
// LFMCMC Getters - N values
// *************************************

[[cpp11::register]]
int get_n_samples_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return static_cast<int>(lfmcmc_ptr->get_n_samples());

}

[[cpp11::register]]
int get_n_stats_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return static_cast<int>(lfmcmc_ptr->get_n_stats());

}

[[cpp11::register]]
int get_n_params_cpp(SEXP lfmcmc) {

    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    return static_cast<int>(lfmcmc_ptr->get_n_params());

}

// *************************************
// LFMCMC Printing functions
// *************************************

[[cpp11::register]]
SEXP verbose_off_lfmcmc_cpp(SEXP lfmcmc) {

  WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
  lfmcmc_ptr->verbose_off();
  return lfmcmc;

}

[[cpp11::register]]
SEXP verbose_on_lfmcmc_cpp(SEXP lfmcmc) {

  WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
  lfmcmc_ptr->verbose_on();
  return lfmcmc;

}

[[cpp11::register]]
SEXP set_params_names_cpp(
    SEXP lfmcmc,
    std::vector< std::string > names
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->set_params_names(names);
    return lfmcmc;
}

[[cpp11::register]]
SEXP set_stats_names_cpp(
    SEXP lfmcmc,
    std::vector< std::string > names
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->set_stats_names(names);
    return lfmcmc;
}

[[cpp11::register]]
SEXP print_lfmcmc_cpp(
    SEXP lfmcmc,
    int burnin
) {
    WrapLFMCMC(lfmcmc_ptr)(lfmcmc);
    lfmcmc_ptr->print(static_cast<size_t>(burnin));
    return lfmcmc;
}

#undef WrapLFMCMC

File: R/LFMCMC.R
#' Likelihood-Free Markhov Chain Monte Carlo (LFMCMC)
#'
#' @aliases epiworld_lfmcmc
#' @param model A model of class [epiworld_model] or `NULL` (see details).
#' @param fun A function (see details).
#' @details
#' Performs a Likelihood-Free Markhov Chain Monte Carlo simulation. When
#' `model` is not `NULL`, the model uses the same random-number generator
#' engine as the model. Otherwise, when `model` is `NULL`, a new random-number
#' generator engine is created.
#'
#' The functions passed to the LFMCMC object have different arguments depending
#' on the object:
#' - `set_proposal_fun`: A vector of parameters and the model.
#' - `set_simulation_fun`: A vector of parameters and the model.
#' - `set_summary_fun`: A vector of simulated data and the model.
#' - `set_kernel_fun`: A vector of simulated statistics, observed statistics,
#' epsilon, and the model.
#'
#' @returns
#' The `LFMCMC` function returns a model of class [epiworld_lfmcmc].
#' @examples
#' ## Setup an SIR model to use in the simulation
#' model_seed <- 122
#' model_sir <- ModelSIR(name = "COVID-19", prevalence = .1,
#'   transmission_rate = .9, recovery_rate = .3)
#' agents_smallworld(
#'   model_sir,
#'   n = 1000,
#'   k = 5,
#'   d = FALSE,
#'   p = 0.01
#' )
#' verbose_off(model_sir)
#' run(model_sir, ndays = 50, seed = model_seed)
#'
#' ## Setup LFMCMC
#' # Extract the observed data from the model
#' obs_data <- get_today_total(model_sir)
#'
#' # Define the simulation function
#' simfun <- function(params, lfmcmc_obj) {
#'   set_param(model_sir, "Recovery rate", params[1])
#'   set_param(model_sir, "Transmission rate", params[2])
#'   run(model_sir, ndays = 50)
#'   res <- get_today_total(model_sir)
#'   return(res)
#' }
#'
#' # Define the summary function
#' sumfun <- function(dat, lfmcmc_obj) {
#'   return(dat)
#' }
#'
#' # Create the LFMCMC model
#' lfmcmc_model <- LFMCMC(model_sir) |>
#'   set_simulation_fun(simfun) |>
#'   set_summary_fun(sumfun) |>
#'   use_proposal_norm_reflective() |>
#'   use_kernel_fun_gaussian() |>
#'   set_observed_data(obs_data)
#'
#' ## Run LFMCMC simulation
#' # Set initial parameters
#' par0 <- c(0.1, 0.5)
#' n_samp <- 2000
#' epsil <- 1.0
#'
#' # Run the LFMCMC simulation
#' verbose_off(lfmcmc_model)
#' run_lfmcmc(
#'   lfmcmc = lfmcmc_model,
#'   params_init = par0,
#'   n_samples = n_samp,
#'   epsilon = epsil,
#'   seed = model_seed
#' )
#'
#' # Print the results
#' set_stats_names(lfmcmc_model, get_states(model_sir))
#' set_params_names(lfmcmc_model, c("Immune recovery", "Infectiousness"))
#'
#' print(lfmcmc_model)
#'
#' get_mean_stats(lfmcmc_model)
#' get_mean_params(lfmcmc_model)
#'
#' @export
#' @concept fmcmc
LFMCMC <- function(model = NULL) {

  if ((length(model) > 0) && !inherits(model, "epiworld_model"))
    stop(
      "model should be of class 'epiworld_model'. It is of class ",
      paste(class(model), collapse = "\", ")
    )

  structure(
    LFMCMC_cpp(model),
    class = c("epiworld_lfmcmc")
  )

}

#' @rdname LFMCMC
#' @param lfmcmc LFMCMC model
#' @param params_init Initial model parameters, treated as double
#' @param n_samples Number of samples, treated as integer
#' @param epsilon Epsilon parameter, treated as double
#' @param seed Random engine seed
#' @returns The simulated model of class [epiworld_lfmcmc].
#' @export
run_lfmcmc <- function(
    lfmcmc, params_init, n_samples, epsilon,
    seed = NULL
    ) {

  stopifnot_lfmcmc(lfmcmc)

  if (length(seed))
    set.seed(seed)

  run_lfmcmc_cpp(
    lfmcmc,
    as.double(params_init),
    as.integer(n_samples),
    as.double(epsilon),
    sample.int(1e4, 1)
  )

  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @param observed_data Observed data, treated as double.
#' @export
set_observed_data <- function(lfmcmc, observed_data) {

  stopifnot_lfmcmc(lfmcmc)

  set_observed_data_cpp(
    lfmcmc,
    as.double(observed_data)
  )

  invisible(lfmcmc)
}

#' @rdname LFMCMC
#' @export
set_proposal_fun <- function(lfmcmc, fun) {

  stopifnot_lfmcmc(lfmcmc)
  set_proposal_fun_cpp(lfmcmc, fun)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @export
use_proposal_norm_reflective <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  use_proposal_norm_reflective_cpp(lfmcmc)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @export
set_simulation_fun <- function(lfmcmc, fun) {

  stopifnot_lfmcmc(lfmcmc)
  set_simulation_fun_cpp(lfmcmc, fun)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @export
set_summary_fun <- function(lfmcmc, fun) {

  stopifnot_lfmcmc(lfmcmc)
  set_summary_fun_cpp(lfmcmc, fun)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @export
set_kernel_fun <- function(lfmcmc, fun) {

  stopifnot_lfmcmc(lfmcmc)
  set_kernel_fun_cpp(lfmcmc, fun)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @returns
#' - `use_kernel_fun_gaussian`: The LFMCMC model with kernel function set to
#' gaussian.
#' @export
use_kernel_fun_gaussian <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  use_kernel_fun_gaussian_cpp(lfmcmc)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @returns
#' - `get_mean_params`: The param means for the given lfmcmc model.
#' @export
get_mean_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_mean_params_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @returns
#' - `get_mean_stats`: The stats means for the given lfmcmc model.
#' @export
get_mean_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_mean_stats_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_initial_params` returns the initial parameters
#' for the given LFMCMC model.
get_initial_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_initial_params_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_current_proposed_params` returns the proposed parameters
#' for the next LFMCMC sample.
get_current_proposed_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_current_proposed_params_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_current_accepted_params` returns the most recently accepted
#' parameters (the current state of the LFMCMC)
get_current_accepted_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_current_accepted_params_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_current_proposed_stats` returns the statistics
#' from the simulation run with the proposed parameters
get_current_proposed_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_current_proposed_stats_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_current_accepted_stats` returns the statistics
#' from the most recently accepted parameters
get_current_accepted_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_current_accepted_stats_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_observed_stats` returns the statistics
#' for the observed data
get_observed_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_observed_stats_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_sample_params` returns a matrix of sample
#' parameters for the given LFMCMC model. with the number of rows equal to the
#' number of samples and the number of columns equal to the number of
#' parameters.
get_all_sample_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  a_params <- get_all_sample_params_cpp(lfmcmc)
  n_params <- get_n_params(lfmcmc)

  matrix(
    a_params,
    ncol = n_params,
    byrow = TRUE
  )

}

#' @export
#' @rdname LFMCMC
#' @returns
#' - The function `get_all_sample_stats` returns a matrix of statistics
#' for the given LFMCMC model. with the number of rows equal to the number of
#' samples and the number of columns equal to the number of statistics.
get_all_sample_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  stats <- get_all_sample_stats_cpp(lfmcmc)
  n_stats <- get_n_stats(lfmcmc)

  matrix(
    stats,
    ncol = n_stats,
    byrow = TRUE
  )

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_sample_acceptance` returns a vector of boolean flags
#' which indicate whether a given sample was accepted
get_all_sample_acceptance <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_all_sample_acceptance_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_sample_drawn_prob` returns a vector of drawn probabilities
#' for each sample
get_all_sample_drawn_prob <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_all_sample_drawn_prob_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_sample_kernel_scores` returns a vector of kernel scores for
#' each sample
get_all_sample_kernel_scores <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_all_sample_kernel_scores_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_accepted_params` returns a matrix of accepted
#' parameters for the given LFMCMC model. with the number of rows equal to the
#' number of samples and the number of columns equal to the number of
#' parameters.
get_all_accepted_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  a_params <- get_all_accepted_params_cpp(lfmcmc)
  n_params <- get_n_params(lfmcmc)

  matrix(
    a_params,
    ncol = n_params,
    byrow = TRUE
  )

}


#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_accepted_stats` returns a matrix of accepted statistics
#' for the given LFMCMC model. with the number of rows equal to the number of
#' samples and the number of columns equal to the number of statistics.
get_all_accepted_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  a_stats <- get_all_accepted_stats_cpp(lfmcmc)
  n_stats <- get_n_stats(lfmcmc)

  matrix(
    a_stats,
    ncol = n_stats,
    byrow = TRUE
  )

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The function `get_all_accepted_kernel_scores` returns a vector of kernel scores for
#' each accepted sample
get_all_accepted_kernel_scores <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_all_accepted_kernel_scores_cpp(lfmcmc)

}

#' @export
#' @rdname LFMCMC
#' @returns
#' - The functions `get_n_samples`, `get_n_stats`, and `get_n_params`
#' return the number of samples, statistics, and parameters for the given
#' LFMCMC model, respectively.
get_n_samples <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_n_samples_cpp(lfmcmc)

}

#' @export
#' @rdname LFMCMC
get_n_stats <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_n_stats_cpp(lfmcmc)

}

#' @export
#' @rdname LFMCMC
get_n_params <- function(lfmcmc) {

  stopifnot_lfmcmc(lfmcmc)
  get_n_params_cpp(lfmcmc)

}

#' @rdname LFMCMC
#' @export
#' @returns
#' - The `verbose_on` and `verbose_off` functions return the same model, however
#' `verbose_off` returns the model with no progress bar.
#' @details
#' The `verbose_on` and `verbose_off` functions activate and deactivate printing
#' progress on screen, respectively. Both functions return the model (`x`) invisibly.

#' @export
verbose_off.epiworld_lfmcmc <- function(x) {
  invisible(verbose_off_lfmcmc_cpp(x))
}

#' @export
verbose_on.epiworld_lfmcmc <- function(x) {
  invisible(verbose_on_lfmcmc_cpp(x))
}

#' @rdname LFMCMC
#' @param names Character vector of names.
#' @returns
#' - `set_params_names`: The lfmcmc model with the parameter names added.
#' @export
set_params_names <- function(lfmcmc, names) {

  stopifnot_lfmcmc(lfmcmc)
  set_params_names_cpp(lfmcmc, names)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @returns
#' - `set_stats_names`: The lfmcmc model with the stats names added.
#' @export
set_stats_names <- function(lfmcmc, names) {

  stopifnot_lfmcmc(lfmcmc)
  set_stats_names_cpp(lfmcmc, names)
  invisible(lfmcmc)

}

#' @rdname LFMCMC
#' @param x LFMCMC model to print
#' @param ... Ignored
#' @param burnin Integer. Number of samples to discard as burnin before
#' computing the summary.
#' @export
print.epiworld_lfmcmc <- function(x, burnin = 0, ...) {

  if (!is.numeric(burnin))
    stop("The 'burnin' argument must be an integer.")

  if (burnin < 0)
    stop("The 'burnin' argument must be a non-negative integer.")

  print_lfmcmc_cpp(x, burnin = burnin)
  invisible(x)

}
